packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,902 @@
+    event ConnectorCallResult (address connector, bool success, bytes reason);
+
+    event ConnectorTrusted    (address connector, bool trusted);
+
+    struct InternalBridgeTransfer
+    {
+        address owner;
+        uint16  tokenID;
+        uint96  amount;
+    }
+
+    struct TokenData
+    {
+        address token;
+        uint16  tokenID;
+        uint    amount;
+    }
+
+    struct ConnectorCalls
+    {
+        address             connector;
+        uint                gasLimit;
+        ConnectorGroup[]    groups;
+    }
+
+    struct TransferBatch
+    {
+        uint     batchID;
+        uint96[] amounts;
+    }
+
+    struct BridgeOperations
+    {
+        TransferBatch[]  transferBatches;
+        ConnectorCalls[] connectorCalls;
+        TokenData[]      tokens;
+    }
+
+    struct Context
+    {
+        TokenData[] tokens;
+        uint        tokensOffset;
+        uint        txsDataPtr;
+        uint        txsDataPtrStart;
+    }
+
+    struct CallTransfer
+    {
+        uint fromAccountID;
+        uint tokenID;
+        uint amount;
+        uint feeTokenID;
+        uint fee;
+        uint storageID;
+        uint packedData;
+    }
+
+    bytes32 constant public BRIDGE_CALL_TYPEHASH = keccak256(
+        "BridgeCall(uint16 tokenID,uint96 amount,uint16 feeTokenID,uint96 maxFee,uint32 validUntil,uint32 storageID,uint32 minGas,address connector,bytes groupData,bytes userData)"
+    );
+
+    uint               public constant  MAX_NUM_TRANSACTIONS_IN_BLOCK = 386;
+    uint               public constant  MAX_AGE_PENDING_TRANSFER      = 7 days;
+    uint               public constant  MAX_FEE_BIPS                  = 25;     // 0.25%
+    uint               public constant  GAS_LIMIT_CHECK_GAS_LIMIT     = 10000;
+
+    IExchangeV3        public immutable exchange;
+    uint32             public immutable accountID;
+    IDepositContract   public immutable depositContract;
+    bytes32            public immutable DOMAIN_SEPARATOR;
+
+    address                                     public exchangeOwner;
+
+    mapping (uint => mapping (bytes32 => uint)) public pendingTransfers;
+    mapping (uint => mapping(uint => bool))     public withdrawn;
+
+    mapping (address => bool)                   public trustedConnectors;
+
+    uint                                        public batchIDGenerator;
+
+    // token -> tokenID
+    mapping (address => uint16)                 public cachedTokenIDs;
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchangeOwner, "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3      _exchange,
+        uint32           _accountID
+        )
+    {
+        exchange = _exchange;
+        accountID = _accountID;
+
+        depositContract = _exchange.getDepositContract();
+        exchangeOwner = _exchange.owner();
+
+        DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("Bridge", "1.0", address(this)));
+    }
+
+    function batchDeposit(
```

Comment:
shall we add `onlyFromExchangeOwner` just as we did for BaseConverter?

Reply:
No, we want this to be callable by everyone! This is the function other rollups/CEXs can use to mass migrate users to Loopring. The relayer doesn't need to know who called it because it's the same format as the batch deposits done by the connectors.

Most of the gas cost is paid by the caller of this function because this already does the actual deposit to the exchange. The only thing left to do after this function is the distribution on L2 which costs ~3,000 gas. To help pay for this we can take `MAX_FEE_BIPS` of the amount deposited. But if we get spammed or this is not enough to cover the costs the relayer can also just ignore the batch deposit request.

Reply:
> Brecht, do we really want this Bridge to provide the batchDeposit method? It seems the internal _batchDeposit is really needed to do the L1->L2 token distribution, but allowing other people to call the external batchDeposit seems not an expected use case.

I would say it's certainly very much an expected use case, I would say it's even one of the 2 main features of the Bridge ( 1) general L1 -> L2 and 2) general L2 -> L1, and that's basically all the bridge does). The fact that the L1 -> L2 token distribution is slightly differen because an internal batch deposit function is only for optimization for trusted connectors, otherwise connectors would also just use the public batchDeposit function (untrusted connectors have to do batch deposits like that).